examples/step-79/step-79.cc

```diff
@@ -0,0 +1,2755 @@
+            for (auto check_cell : cells_to_check)
+              {
+                for (unsigned int n = 0; n < GeometryInfo<dim>::faces_per_cell;
+                     ++n)
+                  {
+                    if (!(check_cell->face(n)->at_boundary()))
+                      {
+                        const double distance = cell->center().distance(
+                          check_cell->neighbor(n)->center());
+                        if ((distance < filter_r) &&
+                            !(neighbor_ids.count(
+                              check_cell->neighbor(n)->active_cell_index())))
+                          {
+                            cells_to_check_temp.insert(check_cell->neighbor(n));
+                            neighbor_ids.insert(
+                              check_cell->neighbor(n)->active_cell_index());
+                          }
+                      }
+                  }
+              }
+
+            if (neighbor_ids.size() == n_neighbors)
+              break;
+            else
+              {
+                cells_to_check = cells_to_check_temp;
+                n_neighbors    = neighbor_ids.size();
+              }
+          }
+
+        for (const auto j : neighbor_ids)
+          {
+            dsp
+              .block(SolutionBlocks::unfiltered_density,
+                     SolutionBlocks::unfiltered_density_multiplier)
+              .add(i, j);
+            dsp
+              .block(SolutionBlocks::unfiltered_density_multiplier,
+                     SolutionBlocks::unfiltered_density)
+              .add(i, j);
+          }
+      }
+
+    // Having so generated the "dynamic" sparsity pattern, we can
+    // finally copy it to the structure that is used to associate
+    // matrices with a sparsity pattern. Because the sparsity
+    // pattern is large and complex, we also output it into a file
+    // of its own for visualization purposes -- in other words,
+    // for "visual debugging".
+    sparsity_pattern.copy_from(dsp);
+
+    std::ofstream out("sparsity.plt");
+    sparsity_pattern.print_gnuplot(out);
+
+    system_matrix.reinit(sparsity_pattern);
+
+
+    // What is left is to correctly size the various vectors and
+    // their blocks, as well as setting initial guesses for some
+    // of the components of the (nonlinear) solution vector. We
+    // here use the symbolic component names for individual blocks
+    // of the solution vector and, for brevity, use the same trick
+    // with `using namespace` as above:
+    nonlinear_solution.reinit(block_sizes);
+    system_rhs.reinit(block_sizes);
+
+    {
+      using namespace SolutionBlocks;
+
+      for (unsigned int k = 0; k < n_u; ++k)
+        {
+          nonlinear_solution.block(displacement)[k]            = 0;
+          nonlinear_solution.block(displacement_multiplier)[k] = 0;
+        }
+      for (unsigned int k = 0; k < n_p; ++k)
+        {
+          nonlinear_solution.block(density)[k]            = density_ratio;
```

Comment:
I think you can replace the loop by `` nonlinear_solution.block(density).add(density_ratio);``, etc.